### DISCUSSION TITLE
Randomization

### DISCUSSION THREAD

**Discussion by ca-watson** [jspsych/jsPsych#1015] [Q&A]
I am having a difficult time figuring out the randomization. I am trying to randomize these 'blocks' and everything works fine but the actual randomization. By means of reading the various jspsych tutorials, it seems correct, but maybe I am missing something. 

  var square1 = {
        timeline:
        [
           {
            type: 'video',
            width: 600,
            sources: ['video/lg01.mp4']
          },
          {
            stimulus: '<p>Given the animation you just saw, .... </p>' 
            type: 'html-slider-response',
            width: 600,
            labels: ['Completely Unlikely', 'Completely Likely'],
            button_label: ['Submit'],
            data: {
              testpart: 'mapping_scale',
              subject: code,
            }
          }
        ]
    };
    timeline.push(fixation);
    timeline.push(square1);

  var square2 = {
      timeline:
      [
        {
          type: 'video',
          width: 600,
          sources: ['video/lg01.mp4']
        },
        {
          stimulus: '<p>Given the animation you just saw, .... </p>' 
          type: 'html-slider-response',
          width: 600,
          labels: ['Completely Unlikely', 'Completely Likely'],
          button_label: ['Submit'],
          data: {
            testpart: 'mapping_scale',
            subject: code,
          }
        }
      ]
  };
  timeline.push(fixation);
  timeline.push(square2);

  var circle1 = {
      timeline:
      [
        {
          type: 'video',
          width: 600,
          sources: ['video/lg02.mp4']
        },
        {
          stimulus: '<p>Given the animation you just saw, .... </p>' 
          type: 'html-slider-response',
          width: 600,
          labels: ['Completely Unlikely', 'Completely Likely'],
          button_label: ['Submit'],
          data: {
            testpart: 'mapping_scale',
            subject: code,
          }
        }
      ]
  };
  timeline.push(fixation);
  timeline.push(circle1);

 var circle2 = {
      timeline:
      [
        {
          type: 'video',
          width: 600,
          sources: ['video/lg02.mp4']
        },
        {
          stimulus: '<p>Given the animation you just saw, .... </p>' 
          type: 'html-slider-response',
          width: 600,
          labels: ['Completely Unlikely', 'Completely Likely'],
          button_label: ['Submit'],
          data: {
            testpart: 'mapping_scale',
            subject: code,
          }
        }
      ]
  };
  timeline.push(fixation);
  timeline.push(circle2);

  var myArray= ['circle1', 'circle2', 'square1', 'square2'];
  var random_order = jsPsych.randomization.sampleWithoutReplacement(myArray, 4);

**Comment by becky-gilbert** [comment]
Just to be sure I understand - I think you want to randomize your 4 different trial timelines (video + slider response), with a fixation cross before each one. Is that right?

If so, one issue is that you're adding each trial timeline to your main timeline after you create it (using `timeline.push(square1)`, etc.). This will put your trial timelines onto the main timeline in a fixed order. Instead, you could use the `jsPsych.randomization.sampleWithoutReplacement()` function to generate a random order, as you've already done, and then use the resulting random_order array to determine the order that your trials are added onto the main timeline. 

Another issue is that I think it's not currently possible to create an object with just a timeline and nothing else (see issue #938). At the moment I think this only works if you're using timeline_variables, conditional_function, or loop_function. So instead of creating objects that contain each trial timeline, you would have to add each trial object (fixation, video, slider) onto the main timeline separately.

I think the easiest solution is to set up your trial timeline with [timeline_variables](https://www.jspsych.org/overview/timeline/#timeline-variables) because you're repeating the same general trial procedure but just with a few different parameters across trials. This will allow you to set up your trial procedure just once (fixation, video, slider). Then you would need a timeline_variables array contains the set of parameter values that changes across trials. That way you can just set [randomize_order](https://www.jspsych.org/overview/timeline/#random-orders-of-trials) to true to randomize your trials.

**Reply by ca-watson** [reply]
This was super helpful! Thank you. (I am very new at coding, so its been rough) 

If you don't mind, I have another question maybe you could help with. here is my new code per the timeline_variable, but i keep getting an error "You must specify a value for the sources parameter in the video plugin." I feel like I am pulling it from the timeline variable 'video' and I cant seem to find an alternative solution. 


var uncertainty = {
    timeline: [
        {
          type: 'html-keyboard-response',
          stimulus: '<div style="font-size: 60px;"> + </div>',
          choices: jsPsych.NO_KEYS,
          trial_duration: 500
        },
        {
          type: 'video',
          width: 600,
          source: jsPsych.timelineVariable('video')
        },
        {
          type: 'html-slider-response',
          width: 600,
          stimulus: jsPsych.timelineVariable('sentence'),
          labels: ['Completely Unlikely', 'Completely Likely'],
          button_label: ['Submit'],
          data: {
            testpart: 'mapping_scale',
            subject: code
          }
        }
      ],
        timeline_variable: [
          {video: 'video/lg01.mp4', sentence: 'Did the square turn blue?'},
          {video: 'video/lg02.mp4', sentence: 'Did the square turn green?'},
          {video: 'video/lg01.mp4', sentence: 'Did the square turn yellow?'}
        ],
        randomize_order: true
      }
        timeline.push(uncertainty);

**Reply by ca-watson** [reply]
The fixation cross works, but then it stops at the video portion. Also, a note i am using jspsych 6.0.5.

**Reply by becky-gilbert** [reply]
I'm not sure about v6.0.5, but in v6.1.0, the video plugins (e.g. [video-keyboard-response](https://www.jspsych.org/plugins/jspsych-video-keyboard-response/#parameters)) take a parameter called `sources` (plural). The value has to be an array, rather than just a single string. This is to allow people to provide different file format options for their video, which improves cross-browser compatibility.

So you could try changing the parameter name to `sources` in your trial object. And in your timeline variables array, change your video files from a file path string to an arrays that contain one or more file paths:
```javascript
{video: ['video/lg01.mp4'], sentence: 'Did the square turn blue?'}
```
Also, there's no plugin called 'video' in jsPsych v6.1.0, so just checking that the plugin name used in the trial's `type` parameter is correct?
